Proposal for opaque log tags

RaduBerinde opened the pull request:
This is a proposal for adding opaque "tags" to contexts that show up in log
messages.

See log_context_test for an example.

@tschottdorf @jordanlewis

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7966)

<!-- Reviewable:end -->

tschottdorf commented on util/log/log_context.go:
```diff
@@ -0,0 +1,56 @@
+}
+
+// Unique handle for the contextTags value (see context.Value).
+var contextTagKey interface{} = new(int)
```
```go
type contextTag struct{}
context.WithValue(ctx, contextTag{}, newTag)
```

tschottdorf commented:
Looks useful to me as one of the "baby steps" @petermattis mentioned in https://github.com/cockroachdb/cockroach/issues/1779.

Assuming nobody turns out to be wildly against this and you take this further, how about removing the non-contextualized version of the logging family (which should be good for future discipline and also to retrofit existing logging which has a context nearby)?

There's also undoing part of my PR #7914 (removal of `.context()`) and some of @arjunravinarayan's `log.Shmogf("%s: ...", replica)` formattings.

petermattis commented:
Looks reasonable to me. 

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[util/log/log_context.go, line 23 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNENoVh-DsKGopkyxQ-:-KNENoVh-DsKGopkyxQ0:745510522) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L23)):*
> ```Go
> type logTag struct {
> 	name  string
> 	value interface{}
> ```

I wonder if we should take a page from `zap` and have specializations for `int` and `string`.

---

*[util/log/log_context.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNEMKL_hR9eXaMedPum:-KNEMKL_hR9eXaMedPun:1867827847) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L43)):*
> ```Go
> 		tags = append(tags, *t)
> 	}
> 	for i, j := 0, len(tags)-1; i < j; i, j = i+1, j-1 {
> ```

`i++, j--`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-:-KNEO-Sq2yCnRZiuAc65:-1424177314)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Thanks!

@tschottdorf I agree that consolidating the functions to always use contexts that is the right thing to do. I would probably remove the `log.Mehc` versions and add a context to the `Meh` and `Mehf` versions). But I'm not sure if it's best for a single person to do it, perhaps we should find a person in each component that is somewhat familiar with the code? I can do it for SQL, my worry is that in a lot of other areas there may be contexts stashed in non-obvious places.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[util/log/log_context.go, line 23 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNENoVh-DsKGopkyxQ-:-KNHEKflrWiFdy8vLn3M:-1375541016) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L23)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I wonder if we should take a page from `zap` and have specializations for `int` and `string`.
</details>
To avoid the allocation when boxing into an `interface{}`? That only works if we provide different functions for each type (otherwise we are still boxing just to pass it to the `WithLogTags`), do you think that's worth it? (By the way, as I recently found out, I don't think that saves anything for `int`s, since they can be embedded in an interface directly).

Somewhat related - I can save an allocation by reimplementing WithValue and embeedding the tag in the context directly (instead of allocating a separate logTag).

---

*[util/log/log_context.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNDcwvBaphR1HfOWYee-r1-28:-KNHEgU5ld_8iTMyJkdM:1296255418) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L28)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> ```go
> 
> type contextTag struct{}
> 
> context.WithValue(ctx, contextTag{}, newTag)
> 
> ```
</details>
I didn't go that route because I thought that this results in an allocation (since the `interface{}` we are passing to `WithValue` has to point to something). But I'm reading that if the target type fits in a word it is embedded in the interface structure directly, so there should be no allocation. I'll switch to this.

---

*[util/log/log_context.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNEMKL_hR9eXaMedPum:-KNHGH892o9BOtZ5Z7QA:469309182) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L43)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `i++, j--`?
</details>
Not in go :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-:-KNHDi2cfKwsrB_fXCxO:-711186304)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[util/log/log_context.go, line 23 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNENoVh-DsKGopkyxQ-:-KNHqcfSZvTNKdfu08YI:-789105373) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L23)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> To avoid the allocation when boxing into an `interface{}`? That only works if we provide different functions for each type (otherwise we are still boxing just to pass it to the `WithLogTags`), do you think that's worth it? (By the way, as I recently found out, I don't think that saves anything for `int`s, since they can be embedded in an interface directly).
> 
> Somewhat related - I can save an allocation by reimplementing WithValue and embeedding the tag in the context directly (instead of allocating a separate logTag).
</details>
Yes, to avoid the allocation when boxing an `interface{}`. Yes, we would need `WithLogTagInt` and `WithLogTagString`. Probably not worth it until we see that this is costly.

It doesn't look possible to reimplement `WithValue` as the `context` package internally uses a type assertion to see if something is a `valueCtx` when propagating cancellation. Perhaps I'm missing something.

---

*[util/log/log_context.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNDcwvBaphR1HfOWYee-r1-28:-KNHqtTxFngxPV5lFNrg:543306030) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I didn't go that route because I thought that this results in an allocation (since the `interface{}` we are passing to `WithValue` has to point to something). But I'm reading that if the target type fits in a word it is embedded in the interface structure directly, so there should be no allocation. I'll switch to this.
</details>
An empty struct consumes no space and can be stored in an `interface{}` without an allocation. Not that only empty structs and pointers have this feature. A word-sized object like an integer requires an allocation because the second word in an `interface{}` must be a pointer.

---

*[util/log/log_context.go, line 43 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNEMKL_hR9eXaMedPum:-KNHr6QOIafns5A8069G:1007889841) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L43)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Not in go :)
</details>
Ugh. I forgot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-:-KNHr9p9LzMJOQ95-bRP:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[util/log/log_context.go, line 23 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNENoVh-DsKGopkyxQ-:-KNIGTo_BzOhMCprkPPO:-212041539) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L23)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yes, to avoid the allocation when boxing an `interface{}`. Yes, we would need `WithLogTagInt` and `WithLogTagString`. Probably not worth it until we see that this is costly.
> 
> It doesn't look possible to reimplement `WithValue` as the `context` package internally uses a type assertion to see if something is a `valueCtx` when propagating cancellation. Perhaps I'm missing something.
</details>
Ah, you're right. That's unfortunate.

---

*[util/log/log_context.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNDcwvBaphR1HfOWYee-r1-28:-KNIFcsJDSg1rOvRwjPB:1581615690) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L28)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> An empty struct consumes no space and can be stored in an `interface{}` without an allocation. Not that only empty structs and pointers have this feature. A word-sized object like an integer requires an allocation because the second word in an `interface{}` must be a pointer.
</details>
Hm.. Ok I need to understand this better.. I will run some experiments.

If the second word needs to be a pointer, how does it work for an empty struct? What's the pointer to in that case?

This document says that anything that fits in a word can be stored directly: http://research.swtch.com/interfaces

But the source code indeed seems to disagree with that document:
https://github.com/golang/go/blob/d25c3eadea9bc5c8b6451c3502d6063dd618a3af/src/cmd/compile/internal/gc/subr.go#L2279
(though from this code it seems that the empty array isn't "direct" afterall?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[util/log/log_context.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNDcwvBaphR1HfOWYee-r1-28:-KNIJTGkxdffFthMsBPS:-1443626947) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Hm.. Ok I need to understand this better.. I will run some experiments.
> 
> If the second word needs to be a pointer, how does it work for an empty struct? What's the pointer to in that case?
> 
> This document says that anything that fits in a word can be stored directly: http://research.swtch.com/interfaces
> 
> But the source code indeed seems to disagree with that document:
> https://github.com/golang/go/blob/d25c3eadea9bc5c8b6451c3502d6063dd618a3af/src/cmd/compile/internal/gc/subr.go#L2279
> (though from this code it seems that the empty array isn't "direct" afterall?)
</details>
Ok, I ran some experiments and what the code above suggests is true. `struct{}` requires an allocation, as does `int`. Pointers, channels, and maps don't. I checked with a small program by generating assembler `go build -gcflags='-S'` and looking for the `runtime.convT2E` call.

So I guess my code stays :)

@nvanbenschoten we really need to work on that go blog post :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[util/log/log_context.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNDcwvBaphR1HfOWYee-r1-28:-KNIR68ubww2AousM_OF:315356135) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Ok, I ran some experiments and what the code above suggests is true. `struct{}` requires an allocation, as does `int`. Pointers, channels, and maps don't. I checked with a small program by generating assembler `go build -gcflags='-S'` and looking for the `runtime.convT2E` call.
> 
> So I guess my code stays :)
> 
> @nvanbenschoten we really need to work on that go blog post :)
</details>
Total speculation on my part but I am guessing that http://research.swtch.com/interfaces was true when it was written, and it changed when the go runtime was written in go (go doesn't allow you to have a "union" and interpret the same data as one of multiple types). Pity..

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[util/log/log_context.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNDcwvBaphR1HfOWYee-r1-28:-KNIfIFYr47S3c66Ltep:1227709078) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Total speculation on my part but I am guessing that http://research.swtch.com/interfaces was true when it was written, and it changed when the go runtime was written in go (go doesn't allow you to have a "union" and interpret the same data as one of multiple types). Pity..
</details>
The `research.swtch.com` document is out of date. I'm 99.9% certain that assigning an empty struct to an interface does not require an allocation. An empty struct does not require any memory and is thus not storing anything in the second word of the interface. Care to share your experiments that suggest that an allocation is happening in that case?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[util/log/log_context.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNDcwvBaphR1HfOWYee-r1-28:-KNIgwWFJBiKs_kDvFFv:1600541131) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L28)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The `research.swtch.com` document is out of date. I'm 99.9% certain that assigning an empty struct to an interface does not require an allocation. An empty struct does not require any memory and is thus not storing anything in the second word of the interface. Care to share your experiments that suggest that an allocation is happening in that case?
</details>
This is from running a test with `-memprofile mem.prof -memprofilerate=1` and then `go tool pprof --alloc_objects`:

```
         .          .      8:func TestAlloc(t *testing.T) {
         .          .      9:	var i interface{}
         .          .     10:	i = struct{}{}
         .          .     11:	fmt.Printf("%v\n", i)
         1          1     12:	i = int(0)
         .          .     13:	fmt.Printf("%v\n", i)
         .          .     14:}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[util/log/log_context.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNDcwvBaphR1HfOWYee-r1-28:-KNIkzNCYolDXBv09lBN:-620403603) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L28)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This is from running a test with `-memprofile mem.prof -memprofilerate=1` and then `go tool pprof --alloc_objects`:
> 
> ```
>          .          .      8:func TestAlloc(t *testing.T) {
>          .          .      9:	var i interface{}
>          .          .     10:	i = struct{}{}
>          .          .     11:	fmt.Printf("%v\n", i)
>          1          1     12:	i = int(0)
>          .          .     13:	fmt.Printf("%v\n", i)
>          .          .     14:}
> ```
</details>
Interesting, thanks for trying it out! What I tried was  https://gist.github.com/anonymous/ea0b2bbfb1de5326723d8a43c7f3da92

And I assumed that calling `runtime.convT2E` means it's not a `isDirectIface` and it will allocate, but maybe I am wrong there https://github.com/golang/go/blob/master/src/runtime/iface.go#L162

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[util/log/log_context.go, line 28 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-KNDcwvBaphR1HfOWYee-r1-28:-KNImNQFUF6eoY-BIkka:-202057915) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d7f8f67a61e7aec4d5c89cf40c6e3879b45e81e/util/log/log_context.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Interesting, thanks for trying it out! What I tried was  https://gist.github.com/anonymous/ea0b2bbfb1de5326723d8a43c7f3da92
> 
> And I assumed that calling `runtime.convT2E` means it's not a `isDirectIface` and it will allocate, but maybe I am wrong there https://github.com/golang/go/blob/master/src/runtime/iface.go#L162
</details>
Ah! I think it is indeed calling `newobject`, but since it's zero that just returns a constant pointer (https://github.com/golang/go/blob/6181db53dbfec513300a236debbdc01735f00c07/src/runtime/malloc.go#L534).

So it's not a "direct interface" - the interface is indeed storing a pointer, it's just that the pointer is fixed and we don't actually need to allocate anything.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Updated. TFTRs and sorry for the golang minutiae detour :)

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-:-KNIvykEhyUZGFXekBQs:-986984219)*
<!-- Sent from Reviewable.io -->

petermattis commented:
No worries about the detour through go minutiae. Good for that knowledge to be disseminated. And I learned about how the zero-size struct allocations were implemented.

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7966#-:-KNJHZzbdQOakoIWZlNL:-1413540340)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
Opaque log tags

This is a proposal for adding opaque "tags" to contexts that show up in log
messages. See log_context_test for an example.

Part of #1779.
